pkg/network/protocols/http/kafka_monitor_test.go

```diff
@@ -37,76 +36,392 @@ const (
+		return fmt.Sprintf("%s-%d", "franz-kafka", testIndex)
+	}
 
-	record := &kgo.Record{Topic: defaultTopicName, Value: []byte("Hello Kafka!")}
-	ctxTimeout, cancel = context.WithTimeout(context.Background(), time.Second*5)
-	err = client.ProduceSync(ctxTimeout, record).FirstErr()
-	cancel()
-	require.NoError(t, err, "record had a produce error while synchronously producing")
-
-	fetches := client.PollFetches(context.Background())
-	errs := fetches.Errors()
-	for _, err := range errs {
-		t.Errorf("PollFetches error: %+v", err)
-		t.FailNow()
+	kafkaTeardown := func(t *testing.T, ctx testContext) {
+		if _, ok := ctx.extras["client"]; !ok {
+			return
+		}
+		client := ctx.extras["client"].(*kafka.Client)
+		defer client.Client.Close()
+		for k, value := range ctx.extras {
+			if strings.HasPrefix(k, "topic_name") {
+				require.NoError(t, client.DeleteTopic(value.(string)))
+			}
+		}
 	}
 
-	// We expect 2 occurrences for each connection as we are working with a docker
-	expectedStatsCount := 4
+	serverAddress := net.JoinHostPort(serverHost, kafkaPort)
+	targetAddress := net.JoinHostPort(targetHost, kafkaPort)
+	kafka.RunServer(t, serverHost, kafkaPort)
+
+	defaultDialer := &net.Dialer{
+		LocalAddr: &net.TCPAddr{
+			IP: net.ParseIP(clientHost),
+		},
+	}
+
+	tests := []kafkaParsingTestAttributes{
+		{
+			name: "Sanity - produce and fetch",
+			context: testContext{
+				serverPort:    kafkaPort,
+				targetAddress: targetAddress,
+				serverAddress: serverAddress,
+				extras: map[string]interface{}{
+					"topic_name": getTopicName(),
+				},
+			},
+			testBody: func(t *testing.T, ctx testContext, monitor *Monitor) {
+				topicName := ctx.extras["topic_name"].(string)
+				client, err := kafka.NewClient(kafka.Options{
+					ServerAddress: ctx.targetAddress,
+					Dialer:        defaultDialer,
+					CustomOptions: []kgo.Opt{
+						kgo.MaxVersions(kversion.V2_5_0()),
+						kgo.ConsumeTopics(topicName),
+					},
+				})
+				ctx.extras["client"] = client
+				require.NoError(t, err)
+				require.NoError(t, client.CreateTopic(ctx.extras["topic_name"].(string)))
+
+				record := &kgo.Record{Topic: topicName, Value: []byte("Hello Kafka!")}
+				ctxTimeout, cancel := context.WithTimeout(context.Background(), time.Second*5)
+				defer cancel()
+				require.NoError(t, client.Client.ProduceSync(ctxTimeout, record).FirstErr(), "record had a produce error while synchronously producing")
+
+				fetches := client.Client.PollFetches(context.Background())
+				errs := fetches.Errors()
+				for _, err := range errs {
+					t.Errorf("PollFetches error: %+v", err)
+					t.FailNow()
+				}
+
+				// We expect 2 occurrences for each connection as we are working with a docker, so (1 produce + 1 fetch) * 2 = (4 stats)
+				kafkaStats := getAndValidateKafkaStats(t, monitor, 4)
+
+				// kgo client is sending an extra fetch request before running the test, so double the expected fetch request
+				validateProduceFetchCount(t, kafkaStats, topicName, 2, 4, 8, 11)
```

Comment:
a lot of numbers in a row, can you document them or use consts/variables?

Reply:
Introduced `kafkaParsingValidation` struct for better test readability